presto-verifier/src/main/java/com/facebook/presto/verifier/framework/DataVerification.java

```diff
@@ -91,23 +102,54 @@ public DataMatchResult verify(
             ChecksumQueryContext controlChecksumQueryContext,
             ChecksumQueryContext testChecksumQueryContext)
     {
-        List<Column> controlColumns = getColumns(getHelperAction(), typeManager, control.getObjectName());
         List<Column> testColumns = getColumns(getHelperAction(), typeManager, test.getObjectName());
-
-        Query controlChecksumQuery = checksumValidator.generateChecksumQuery(control.getObjectName(), controlColumns);
         Query testChecksumQuery = checksumValidator.generateChecksumQuery(test.getObjectName(), testColumns);
-
-        controlChecksumQueryContext.setChecksumQuery(formatSql(controlChecksumQuery));
         testChecksumQueryContext.setChecksumQuery(formatSql(testChecksumQuery));
 
-        QueryResult<ChecksumResult> controlChecksum = callAndConsume(
-                () -> getHelperAction().execute(controlChecksumQuery, CONTROL_CHECKSUM, ChecksumResult::fromResultSet),
-                stats -> stats.getQueryStats().map(QueryStats::getQueryId).ifPresent(controlChecksumQueryContext::setChecksumQueryId));
+        List<Column> controlColumns = null;
+        ChecksumResult controlChecksumResult = null;
+
+        if (!skipControl) {
+            controlColumns = getColumns(getHelperAction(), typeManager, control.getObjectName());
+            Query controlChecksumQuery = checksumValidator.generateChecksumQuery(control.getObjectName(), controlColumns);
+            controlChecksumQueryContext.setChecksumQuery(formatSql(controlChecksumQuery));
+
+            QueryResult<ChecksumResult> controlChecksum = callAndConsume(
+                    () -> getHelperAction().execute(controlChecksumQuery, CONTROL_CHECKSUM, ChecksumResult::fromResultSet),
+                    stats -> stats.getQueryStats().map(QueryStats::getQueryId).ifPresent(controlChecksumQueryContext::setChecksumQueryId));
+            controlChecksumResult = getOnlyElement(controlChecksum.getResults());
+        }
+        else if (!saveSnapshot) {
```

Comment:
If it includes the normal run, why would I want to look up a snapshot and then if it is not found it will throw SNAPSHOT_DOES_NOT_EXIST error.

The control and test flow will go through the first if condition because we are not skipping control in a normal run.
The only way we enter this code block for control and test flow is when we want to skip control and then just execute the test alone. For that why do we need to fetch the controlChecksumResult from a snapshot ? Can you help me understand the code flow for the normal run a bit more .

Reply:
Sorry, my previous comment is wrong and confusing.

there are 4 cases to be considered here:

1. Regular run with `skipControl = false`
`if (!skipControl)` block is executed and `else if` block is skipped. That is expected behaviour.
2. Regular run with `skipControl = true`
Both `if (!skipControl)` block and `else if` block is skipped and the `testChecksum` is calculated after the `if else if` statement. That is expected behaviour.
3. query-bank && saveSnapshot == false
only `else if (!saveSnapshot)` block is executed. That is expected behaviour.
4. query-bank && saveSnapshot == true
Both `if (!skipControl)` block and `else if` block is skipped, the `testChecksum` is calculated after the `if else if` statement and snapshot is saved. That is expected behaviour.

The original else if is `else if (QUERY_BANK_MODE && !saveSnapshot)`. It can be simplified to `else if (!saveSnapshot)`.

Reply:
Thanks @v-jizhang  for breaking down the 4 cases.
2 is where I am not convinced.

1 ✔️

2  ❌
> else if block is skipped 

`saveSnapshot` should be false by default.  
See https://github.com/prestodb/presto/blob/751f209230eb1c9f6b6e6ffccb784a9e41f5f962/presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerifierConfig.java#L61
so `else if (!saveSnapshot)` is true and the block will be executed.

3. ✔️

4. ✔️

Reply:
Just ignore case 2 because it does not really exist. Please see the calling code [here](https://github.com/prestodb/presto/pull/19253/files#diff-cef48d5d0467cfcce66f40d59d9a028d38b8096fafa1cbcd73a9ec9fded3c91eR247) and [here](https://github.com/prestodb/presto/pull/19253/files#diff-cef48d5d0467cfcce66f40d59d9a028d38b8096fafa1cbcd73a9ec9fded3c91eR251). 
So `if (!skipControl)` means regular run with `skipControl = false` and `else if (!saveSnapshot)` means `QUERY_BANK_MODE` with `saveSnapshot = false`;

Reply:
The method `verify()` is called from `AbstractVerification.run()` with if statements:
`if (QUERY_BANK_MODE.equals(runningMode) && !skipChecksum)` // For `query-bank` mode
`else if (!skipControl && !skipChecksum)`  // regular mode

So within `verify()`, `if (!skipControl)` means regular run and `else if (!saveSnapshot)` means `query-bank `run.

This block code can't be executed with` skip-control = true`.